sihl/http/request.ml

```diff
@@ -0,0 +1,374 @@
+;;
+
+let replace_or_add_to_list ~f to_add l =
+  let found = ref false in
+  let rec aux acc l =
+    match l with
+    | [] -> if not !found then to_add :: acc |> List.rev else List.rev acc
+    | el :: rest ->
+      if f el to_add
+      then (
+        found := true;
+        aux (to_add :: acc) rest)
+      else aux (el :: acc) rest
+  in
+  aux [] l
+;;
+
+let add_cookie ?sign_with (k, v) t =
+  let cookies = cookies t in
+  let cookies =
+    replace_or_add_to_list
+      ~f:(fun (k2, _v2) _ -> String.equal k k2)
+      ( k
+      , match sign_with with
+        | Some signer -> Cookie.Signer.sign signer v
+        | None -> v )
+      cookies
+  in
+  let cookie_header = cookies |> List.map Cookie.make |> Cookie.to_cookie_header in
+  add_header_or_replace cookie_header t
+;;
+
+let add_cookie_unless_exists ?sign_with (k, v) t =
+  let cookies = cookies t in
+  if List.exists (fun (k2, _v2) -> String.equal k2 k) cookies
+  then t
+  else add_cookie ?sign_with (k, v) t
+;;
+
+let remove_cookie key t =
+  let cookie_header =
+    cookies t
+    |> List.filter (fun (k, _) -> not (String.equal k key))
+    |> List.map Cookie.make
+    |> Cookie.to_cookie_header
+  in
+  add_header_or_replace cookie_header t
+;;
+
+let content_type t = header "content-type" t
+let set_content_type s t = add_header_or_replace ("content-type", s) t
+
+let find_in_query key query =
+  query
+  |> List.find_opt (fun (k, _) -> k = key)
+  |> Option.map (fun (_, r) -> r)
+  |> fun opt ->
+  Option.bind opt (fun x ->
+      try Some (List.hd x) with
+      | Not_found -> None)
+;;
+
+let urlencoded key t =
+  let open Lwt.Syntax in
+  let* query = to_urlencoded t in
+  Lwt.return @@ find_in_query key query
+;;
+
+let urlencoded_exn key t =
+  let open Lwt.Syntax in
+  let+ o = urlencoded key t in
+  Option.get o
+;;
+
+let query_list t = t.request |> Cohttp.Request.uri |> Uri.query
+let query key t = query_list t |> find_in_query key
+let query_exn key t = query key t |> Option.get
+
+(* TODO [jerben] expose as core middleware *)
+let id req = Opium_kernel.Hmap.find id_key (Opium_kernel.Request.env req)
+
+let to_ctx req =
+  match id req with
+  | Some id -> Core.Ctx.create ~id ()
+  | None -> Core.Ctx.create ()
+;;
+
+let is_get req =
+  match Opium_kernel.Rock.Request.meth req with
+  | `GET -> true
+  | _ -> false
+;;
+
+let contains substring string =
+  let re = Str.regexp_string string in
+  try
+    ignore (Str.search_forward re substring 0);
+    true
+  with
+  | Not_found -> false
+;;
+
+let accepts_html req =
+  Cohttp.Header.get (Opium.Std.Request.headers req) "Accept"
+  |> Option.map (contains "text/html")
+  |> Option.value ~default:false
+;;
+
+let authorization_token t =
+  let ( let* ) = Option.bind in
+  (* TODO [jerben] make this more robust *)
+  let* header = header "authorization" t in
+  match String.split_on_char ' ' header with
+  | [ _; token ] -> Some token
+  | _ -> None
+;;
+
+let param req key =
```

Comment:
Check if we need these 5 params

Reply:
Fixed.

Reply:
I leave it for now, added proper documentation.